zaza/openstack/charm_tests/ovn/setup.py

```diff
@@ -0,0 +1,130 @@
+# Copyright 2020 Canonical Ltd.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Code for configuring OVN tests."""
+
+import logging
+
+import zaza
+
+import zaza.openstack.charm_tests.test_utils as test_utils
+
+
+class _OVNSetupHelper(test_utils.BaseCharmTest):
+    """Helper class to get at the common `config_change` helper."""
+
+    @staticmethod
+    def _get_instance_mtu_from_global_physnet_mtu():
+        """Calculate instance mtu from Neutron API global-physnet-mtu.
+
+        :returns: Value for instance mtu after migration.
+        :rtype: int
+        """
+        n_api_config = zaza.model.get_application_config('neutron-api')
+
+        # NOTE: we would have to adjust this calculation if we use IPv6 tunnel
+        # endpoints
+        GENEVE_ENCAP_OVERHEAD = 38
+        IP4_HEADER_SIZE = 20
+        return int(n_api_config['global-physnet-mtu']['value']) - (
+            GENEVE_ENCAP_OVERHEAD + IP4_HEADER_SIZE)
+
+    def configure_ngw_novs(self):
+        """Configure n-ovs and n-gw units."""
+        cfg = {
+            # To be able to successfully clean up after the Neutron agents we
+            # need to use the 'openvswitch' `firewall-driver`.
+            'firewall-driver': 'openvswitch',
+            # To be able to have instances successfully survive the migration
+            # without communication issues we need to lower the MTU announced
+            # to instances prior to migration.
+            #
+            # NOTE: In a real world scenario the end user would configure the
+            # MTU at least 24 hrs prior to doing the migration to allow
+            # instances to reconfigure as they renew the DHCP lease.
+            #
+            # NOTE: For classic n-gw topologies the `instance-mtu` config
+            # is a NOOP on neutron-openvswitch units, but that is ok.
+            'instance-mtu': self._get_instance_mtu_from_global_physnet_mtu()
+        }
+        apps = ('neutron-gateway', 'neutron-openvswitch')
+        for app in apps:
+            try:
+                zaza.model.get_application(app)
+                for k, v in cfg.items():
+                    logging.info('Setting `{}` to "{}" on "{}"...'
+                                 .format(k, v, app))
+                with self.config_change(cfg, cfg, app):
+                    continue
```

Comment:
I don't understand the intent here. `self.config_change()` is a context manager that yields after having changed the config and before having restored it. So in between we will now execute `continue`. Will that skip the restoring? Or will that be like saying `pass`? 

I decided to try it out:

```
@contextlib.contextmanager
def config_changed():
    print('changing config')
    yield
    print('restoring config back')

for i in (1, 2):
    with config_changed():
        continue
```

results in

```
changing config
restoring config back
changing config
restoring config back
```

So in fact here `continue` behaves like `pass`. I myself learned something new :) But this wasn't your intent either, correct?

Reply:
The intent is to change the configuration and not change it back and that is accomplished by passing in the same Dict to both the default and alternate parameters. And the code does indeed work as you see on the results.

The reason for using the `config_change` helper to do this is that it already deals with the situation where the config already being set etc. which makes this work regardless of what configuration is passed in in the test bundle.

Reply:
sorry I had something else wrong in mind (some mix-up), so the point I raised didn't make much sense I realize now. Thanks for your patience :)

Reply:
No worries I welcome the scrutiny, and the fact that it is not obvious to the reviewer may call for me adding a comment. Your assumption will probably be shared by anyone else looking at that code as the normal use for the helper is to change and change back.